src/NuGet.Clients/NuGet.PackageManagement.UI/ViewModels/PackageItemViewModel.cs

```diff
@@ -814,7 +814,7 @@ private async System.Threading.Tasks.Task ReloadPackageVersionsAsync()
                 LatestVersion = result;
                 Status = GetPackageStatus(LatestVersion, InstalledVersion, AutoReferenced);
             }
-            catch (OperationCanceledException) when (cancellationToken.IsCancellationRequested)
+            catch (OperationCanceledException e) when (e.CancellationToken.IsCancellationRequested || cancellationToken.IsCancellationRequested)
```

Comment:
Isn't `e.CancellationToken.IsCancellationRequested` always true? How can it be false?
Would I have to `throw new CancellationException(tokenWhichIsNotCancelled)` ?  Not understanding how this would happen.

Should the `when` just be removed? What cases are we NOT handling cancellation?

Reply:
I think in most cases it is, but when we're swapping tokens it seemed like it wouldn't always be the case.

There are two cancellations here, user cancellation and timeout.  I guess we still want to know when a timeout occurred but don't care about user cancellations.  This `catch` is trying to ignore the second one there.  From telemetry, the majority of reported faults are coming from the Semaphore throwing due to cancellation

![image](https://github.com/user-attachments/assets/16ee4c88-9d4e-4587-b9ba-226871a9b948)

Under the debugger, this change did the trick.

Reply:
`e.CancellationToken` will always be the cancellation token that triggered the cancellation. So, when it's an HTTP timeout, not NuGet requesting cancellation on its own cancellation token source, then `e.CancellationToken` will be HttpClient's internal cancellation token.  So, I agree with Donnie, this is effectively `when (true)`, or a catch without a `when` clause

Reply:
Also, the point of line 800 is to capture the `_cancellationTokenSource.Token` before `_cancellationTokenSource` gets swapped (it's done before calling `GetVersionsAsync`), so I would expect this to always be the pre-swapped token.  Is that not working as expected?

Reply:
When does the swap happen in the VM? It looks like the VM creates its own source on construction and the only time it's cancelled is when the object is disposed.

Reply:
I've reworked all of this based on my findings while debugging the code.  Our code does not throw an `OperationCanceledException` and instead wraps this outcome as a `TimeoutException`.